File: dsa/stack/stack.cpp
/* TOPIC: Stack Implementation */

// REFERENCE: https://youtu.be/_6COl6V6mng

#include <bits/stdc++.h>
using namespace std;

class Stack {

    public:
        int *arr;
        int top;
        int size;

    // constructor
    Stack(int size) {
        this->size = size;
        this->arr = new int[size];
        this->top = -1;
    }

    // TIME COMPLEXITY: O(1)
    void push(int val) {
        if (this->top == this->size-1) {
            cout<<"stack overflow"<<endl;
            return;
        }
        this->top++;
        this->arr[top] = val;
        return;
    }

    // TIME COMPLEXITY: O(1)
    void pop() {
        if (this->top == -1) {
            cout<<"stack is empty"<<endl;
            return;
        }
        int temp = this->arr[this->top];
        this->top--;
        return;
    }

    // TIME COMPLEXITY: O(1)
    int peek() {
        if (this->top == -1) {
            cout<<"stack is empty"<<endl;
            return -1;
        }
        int temp = this->arr[this->top];
        return temp;
    }

    // TIME COMPLEXITY: O(1)
    bool isEmpty() {
        return this->top == -1;
    }

};

int main() {

    Stack st(5);

    // st.push(1);
    // st.push(2);
    // st.push(3);
    // st.push(4);
    // st.push(5);
    // st.push(6);

    st.pop();

    cout << st.peek() << endl;

    cout << st.isEmpty() << endl;

    return 0;
}

File: dsa/stack/stack.py
""" Stack with size limit using list"""

class Stack:
    def __init__(self, max):
        self.max = max
        self.list = []

    # display
    def display(self):
        print(self.list)

    # check if stack is empty
    def isEmpty(self):
        if self.list == []:
            return True
        else:
            return False

    # check if stack is full
    def isFull(self):
        if len(self.list) == self.max:
            return True
        else:
            return False

    # push
    def push(self, value):
        if self.isFull():
            print("Stack Overflow")
        else:
            self.list.append(value)

    # pop
    def pop(self):
        if self.isEmpty():
            print("Stack Underflow")
        else:
            return self.list.pop()

    # peek
    def peek(self):
        if self.isEmpty():
            print("Stack Underflow")
        else:
            return self.list[len(self.list)-1]

    # delete stack
    def delete(self):
        self.list = None


s = Stack(5)
print(s.isEmpty())

s.push(1)
s.push(2)
s.push(3)
s.push(4)
s.push(5)
s.display()

print(s.isFull())

s.pop()
s.display()

s.delete()
